src/derivative_operators/abstract_operator_functions.jl

```diff
@@ -0,0 +1,211 @@
+        return 0
+    else
+        return A.stencil_coefs[mid + j - i]
+    end
+end
+
+# scalar - colon - colon
+@inline getindex(A::AbstractDerivativeOperator, kr::Colon, jr::Colon) = full(A)
+
+@inline function getindex(A::AbstractDerivativeOperator, rc::Colon, j)
+    T = eltype(A.stencil_coefs)
+    v = zeros(T, A.dimension)
+    v[j] = one(T)
+    copy!(v, A*v)
+    return v
+end
+
+
+# symmetric right now
+@inline function getindex(A::AbstractDerivativeOperator, i, cc::Colon)
+    T = eltype(A.stencil_coefs)
+    v = zeros(T, A.dimension)
+    v[i] = one(T)
+    copy!(v, A*v)
+    return v
+end
+
+
+# UnitRanges
+@inline function getindex(A::AbstractDerivativeOperator, rng::UnitRange{Int}, cc::Colon)
+    m = full(A)
+    return m[rng, cc]
+end
+
+
+@inline function getindex(A::AbstractDerivativeOperator, rc::Colon, rng::UnitRange{Int})
+    m = full(A)
+    return m[rnd, cc]
+end
+
+@inline function getindex(A::AbstractDerivativeOperator, r::Int, rng::UnitRange{Int})
+    m = A[r, :]
+    return m[rng]
+end
+
+
+@inline function getindex(A::AbstractDerivativeOperator, rng::UnitRange{Int}, c::Int)
+    m = A[:, c]
+    return m[rng]
+end
+
+
+@inline function getindex(A::AbstractDerivativeOperator{T}, rng::UnitRange{Int}, cng::UnitRange{Int}) where T
+    N = A.dimension
+    if (rng[end] - rng[1]) > ((cng[end] - cng[1]))
+        mat = zeros(T, (N, length(cng)))
+        v = zeros(T, N)
+        for i = cng
+            v[i] = one(T)
+            #=
+                calculating the effect on a unit vector to get the matrix of transformation
+                to get the vector in the new vector space.
+            =#
+            A_mul_B!(view(mat, :, i - cng[1] + 1), A, v)
+            v[i] = zero(T)
+        end
+        return mat[rng, :]
+
+    else
+        mat = zeros(T, (length(rng), N))
+        v = zeros(T, N)
+        for i = rng
+            v[i] = one(T)
+            #=
+                calculating the effect on a unit vector to get the matrix of transformation
+                to get the vector in the new vector space.
+            =#
+            A_mul_B!(view(mat, i - rng[1] + 1, :), A, v)
+            v[i] = zero(T)
+        end
+        return mat[:, cng]
+    end
+end
+
+#=
+    This the basic A_Mul_B! function which is broken up into 3 parts ie. handling the left
+    boundary, handling the interior and handling the right boundary. Finally the vector is
+    scaled appropriately according to the derivative order and the degree of discretizaiton.
+    We also update the time stamp of the DerivativeOperator inside to manage time dependent
+    boundary conditions.
+=#
+function Base.A_mul_B!(x_temp::AbstractVector{T}, A::Union{DerivativeOperator{T},UpwindOperator{T}}, x::AbstractVector{T}) where T<:Real
+    convolve_BC_left!(x_temp, x, A)
+    convolve_interior!(x_temp, x, A)
+    convolve_BC_right!(x_temp, x, A)
+    scale!(x_temp, 1./(A.dx^A.derivative_order))
+    A.t[] += 1 # incrementing the internal time stamp
```

Comment:
Time should be updated in `update_coefficients!` directly since many algorithms need to be more careful of this, like when being used in adaptive time stepping algorithms.